components/prism-log.js

```diff
@@ -12,27 +12,31 @@ Prism.languages.log = {
 
 	'level': [
 		{
-			pattern: /\b(?:ALERT|CRIT|CRITICAL|EMERG|EMERGENCY|ERR|ERROR|FAILURE|FATAL|SEVERE)\b/,
+			pattern: /\b(?:ERR|ALERT|CRIT|CRITICAL|EMERG|EMERGENCY|ERR|ERROR|FAILURE|FATAL|SEVERE)\b/,
 			alias: ['error', 'important']
 		},
 		{
-			pattern: /\b(?:WARN|WARNING)\b/,
+			pattern: /\b(?:WRN|WARN|WARNING)\b/,
 			alias: ['warning', 'important']
 		},
 		{
-			pattern: /\b(?:DISPLAY|INFO|NOTICE|STATUS)\b/,
+			pattern: /\b(?:INF|DISPLAY|INFO|NOTICE|STATUS)\b/,
 			alias: ['info', 'keyword']
 		},
 		{
-			pattern: /\b(?:DEBUG|FINE)\b/,
+			pattern: /\b(?:DBG|DEBUG|FINE)\b/,
 			alias: ['debug', 'keyword']
 		},
 		{
-			pattern: /\b(?:FINER|FINEST|TRACE|VERBOSE)\b/,
+			pattern: /\b(?:TRC|VRB|FINER|FINEST|TRACE|VERBOSE)\b/,
 			alias: ['trace', 'comment']
 		}
 	],
 
+	'exception': {
+		pattern: /(.*EXCEPTION.*)(\n\s+at\s+.+)+/im
```

Comment:
Welcome to the world of exponential backtracking! I will be your guide.

But seriously: this pattern has an exponential worst-case runtime. What kind of log is this pattern supposed to match? Could you give me an example?

Reply:
I knew this rule would be a problem 😆 It was mainly just to start a discussion. Lots of program logs contain multiline exceptions for when things really go wrong, and my goal here was to highlight the *entire* exception and really make it stand out.

Reply:
I'm on mobile at the moment but you can see an example exception in the logs of [this](https://github.com/jellyfin/jellyfin/issues/2355) issue.

Reply:
What is the benefit of capturing the whole exception? The current highlighting looks ok (not perfect but ok) to me.

![image](https://user-images.githubusercontent.com/20878432/113864115-ca590c00-97aa-11eb-9ca3-9ed6e8130828.png)

Reply:
It's definitely fine right now, but I believe that exception is highlighted with `property` instead of a unique token for errors or exceptions, which would be nice. The multiline check was specifically just to wrap the entire section for highlighting.

Reply:
> a unique token for errors or exceptions, which would be nice

That is not possible without assuming a stack trace format. This error-message-plus-stack-trace format with `at` is somewhat shared by C#, Java, and JS but there are other formats as well like the ones used in C++, Rust, or Python. It's not necessarily a problem that we miss some languages tho.

Also, is it even useful to tokenize the entire error?

Reply:
I think it's useful, but since the performance issues are pretty clear, I will concede that it's probably not worth adding to this ruleset.

Reply:
Performance isn't an issue. For every exponentially backtracking regex, we can find an equivalent fast regex. Don't worry, I'll help you find it :)

Now that I think about it, if we could reliably find stack trace and identify the programming language (Java/JS/C#/etc.), then we could use our JavaStackTrace and JS Stack trace language to highlight them. That would be pretty cool.

For now, could you test on your side whether the following works:

```js
	'exception': {
		pattern: /(^|[^\w.])[\w.]*(?:Exception|Error):.*(?:(?:\r\n?|\n)[ \t]*(?:at[ \t].+|\.{3}.*|Caused by:.*))+/,
		greedy: true,
	},
```

I only tested it for Java stack traces and JS stack traces but it should work for C# stack traces as well.

Reply:
This works relatively well, but still trips up on the following section.

```
System.ArgumentNullException: Value cannot be null. (Parameter 'item')
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|24_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
```

Reply:
I didn't know C# stack traces had this feature... Here's an updated version of the above pattern:

```
	'exception': {
		pattern: /(^|[^\w.])[\w.]*(?:Exception|Error):.*(?:(?:\r\n?|\n)[ \t]*(?:at[ \t].+|\.{3}.*|Caused by:.*|--- .*))+/,
		greedy: true,
	},
```

Reply:
That did the trick, should I add this to the ruleset?

Reply:
Sure! Isn't that what you wanted in the first place?